Include/internal/pycore_stackref.h

```diff
@@ -8,186 +8,150 @@ extern "C" {
 static inline _PyStackRef
-_PyStackRef_StealRef(PyObject *obj)
+_PyStackRef_FromPyObjectSteal(PyObject *obj)
 {
     // Make sure we don't take an already tagged value.
-    assert(((uintptr_t)obj & Py_TAG_DEFERRED) == 0);
-    return ((_PyStackRef){.bits = ((uintptr_t)(obj))});
+    assert(((uintptr_t)obj & Py_TAG) == 0);
+    int tag = (obj == NULL || _Py_IsImmortal(obj)) ? (Py_TAG_DEFERRED) : Py_TAG_PTR;
+    return ((_PyStackRef){.bits = ((uintptr_t)(obj)) | tag});
 }
-#   define PyStackRef_StealRef(obj) _PyStackRef_StealRef(_PyObject_CAST(obj))
-#else
-#   define PyStackRef_StealRef(obj) ((_PyStackRef){.bits = ((uintptr_t)(obj))})
-#endif
+#define PyStackRef_FromPyObjectSteal(obj) _PyStackRef_FromPyObjectSteal(_PyObject_CAST(obj))
 
 // Converts a PyObject * to a PyStackRef, with a new reference
-#if defined(Py_GIL_DISABLED)
 static inline _PyStackRef
-_PyStackRef_NewRefDeferred(PyObject *obj)
+PyStackRef_FromPyObjectNew(PyObject *obj)
 {
     // Make sure we don't take an already tagged value.
-    assert(((uintptr_t)obj & Py_TAG_DEFERRED) == 0);
+    assert(((uintptr_t)obj & Py_TAG) == 0);
     assert(obj != NULL);
-    if (_PyObject_HasDeferredRefcount(obj)) {
+    // TODO (gh-117139): Add deferred objects later.
+    if (_Py_IsImmortal(obj)) {
         return (_PyStackRef){ .bits = (uintptr_t)obj | Py_TAG_DEFERRED };
     }
     else {
-        return (_PyStackRef){ .bits = (uintptr_t)Py_NewRef(obj) };
+        return (_PyStackRef){ .bits = (uintptr_t)(Py_NewRef(obj)) | Py_TAG_PTR };
     }
 }
-#   define PyStackRef_NewRefDeferred(obj) _PyStackRef_NewRefDeferred(_PyObject_CAST(obj))
-#else
-#   define PyStackRef_NewRefDeferred(obj) PyStackRef_NewRef(((_PyStackRef){.bits = ((uintptr_t)(obj))}))
-#endif
+#define PyStackRef_FromPyObjectNew(obj) PyStackRef_FromPyObjectNew(_PyObject_CAST(obj))
 
-#if defined(Py_GIL_DISABLED)
-static inline _PyStackRef
-_PyStackRef_XNewRefDeferred(PyObject *obj)
-{
-    // Make sure we don't take an already tagged value.
-    assert(((uintptr_t)obj & Py_TAG_DEFERRED) == 0);
-    if (obj == NULL) {
-        return Py_STACKREF_NULL;
-    }
-    return _PyStackRef_NewRefDeferred(obj);
-}
-#   define PyStackRef_XNewRefDeferred(obj) _PyStackRef_XNewRefDeferred(_PyObject_CAST(obj))
-#else
-#   define PyStackRef_XNewRefDeferred(obj) PyStackRef_XNewRef(((_PyStackRef){.bits = ((uintptr_t)(obj))}))
-#endif
 
-// Converts a PyStackRef back to a PyObject *.
-#if defined(Py_GIL_DISABLED)
+// Converts a PyStackRef back to a PyObject *, converting deferred references
+// to new references.
 static inline PyObject *
-PyStackRef_StealObject(_PyStackRef tagged)
+PyStackRef_AsPyObjectNew(_PyStackRef tagged)
 {
-    if ((tagged.bits & Py_TAG_DEFERRED) == Py_TAG_DEFERRED) {
-        assert(_PyObject_HasDeferredRefcount(PyStackRef_Get(tagged)));
-        return Py_NewRef(PyStackRef_Get(tagged));
+    if (!PyStackRef_IsNull(tagged) && PyStackRef_IsDeferred(tagged)) {
+        assert(PyStackRef_IsNull(tagged) ||
```

Comment:
The `PyStackRef_IsNull(tagged)` case in the assertion is impossible given the condition in the `if (!PyStackRef_IsNull(tagged) ...`.